### ISSUE TITLE
Blazor Custom Route Constraints

### ISSUE THREAD

**Issue by GravlLift** [dotnet/aspnetcore#28938]
I am trying to use custom route constraint in Blazor, similar to how it currently works for both [razor pages and MVC routes](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-5.0#custom-route-constraints):

```cs
services.AddRouting(options =>
{
    options.ConstraintMap.Add("customName", typeof(MyCustomConstraint));
});
```

... but it appears that blazor route constraints do not consider any custom constraints that have been registered:

https://github.com/dotnet/aspnetcore/blob/af7c0cc2a2f944811041b55f9b4e0ae6cc948fd7/src/Components/Components/src/Routing/RouteConstraint.cs#L57-L85

So I instead encounter the ArgumentException, "Unsupported constraint '{constraint}' in route '{template}'.".

It would be ideal if `RouteConstraint` could be modified to reference `RouteOptions.ContraintMap` as the other routing engines do. Alternatively, needing to use some Blazor-specific registration method to register constraints only for Blazor would also be acceptable.

**Comment by javiercampos** [comment]
Extensible route constraints should be a must (and I thought that was already solved in asp.net ~12/13 years ago). 
The `SetParametersAsync` parsing "hack" might work for some very contrived scenarios... but it completely disallows generalization, and even forbids doing certain scenarios without many more hacks...

Just imagine some route like: `/users/{id}/{action:createOrEditConstraint}` where you want one page to resolve both `/create` and `/edit`, but you want other different page to resolve `/list` (or others). Since you can't have user-defined constraints, `/users/{id}/{action}` matches both pages and you'll get an error.

Yes, it could be solved by creating -yet another component- and either redirect or include the `create/edit` component and `list` component in it, but that's yet another hack, and a ton more of boilerplate (one more component, with all its initialization and parsing per "CRUD" entity in a project).

One more scenario (which I used in razor pages/webapi/MVC a lot) is having short guids as parameters (while internally using `Guid`). Yes, you can just have a string and parse it on `SetParametersAsync`, but you can't generalize that (on a base class, for example), because you wouldn't have access to the inherited class parameter, so you basically need to specify the `SetParameterAsync` method and parse specifically on every single routable view with a short guid parameter (and again, since you can't constraint it -with a regex or something-, you can't distinguish it from other string routes).

Again, nothing that can't be "hacked away", but this is basic, and I'm sure many people (even those not on this thread) would miss this, so I believe `affected-few` doesn't look like a correct label here, to me.

**Comment by Tragen** [comment]
After 2 hours of debugging why my ulong constraint isn't working, I stumbled upon this bug.
For me it looks like a very basic feature. Please fix this soon.
Also ulong and uint should be added to the default constraints.

**Comment by Alerinos** [comment]
I believe this should be introduced in ASP.NET Core 7

**Comment by Alerinos** [comment]
@pranavkm I would like to remind you of this, it is quite an important and useful thing.

**Comment by kyletinsley** [comment]
Regex route constraints would be wonderful.  Trying to constrain a string route parameter to a finite list of string values and want to 404 otherwise.

**Comment by haefele** [comment]
This would be really really useful to have, for example when I want to allow just a couple string values (a enum basically).
Right now it's very tedious and annoying code to write.

**Comment by kcabral817** [comment]
It appears that folks are tone deaf.  This is a requested requirement and now it appears not even 'short' is supported.  I fail to understand why foundational issues like this keep getting put to the back burner.

**Comment by Tragen** [comment]
This ticket is nearly 3 years old and a very basic feature. I wouldn't bet they will fix this anytime soon.

**Comment by StephenWBertrand** [comment]
yep for sure would like this feature.  Looking to be able to have locale codes embedded into urls as way to support direct links for desired languages etc.

domain.com/{locale?}/myPage
domain.com/myPage  
domain.com/en-US/myPage
domain.com/es-US/myPage

**Comment by ax-meyer** [comment]
I just spent a few hours trying to figure out why my constraint doesn't work... at least now I know.

But please, support such a basic feature. I am very suprised it is not supported for blazor.

**Comment by AndreyZ1** [comment]
here is working solution for ulong data type, works fine:

Route constraint:
```
public class UlongRouteConstraint : IRouteConstraint
{
    public static string UlongRouteConstraintName = "ulong";

    public bool Match(HttpContext? httpContext, IRouter? route, string routeKey, RouteValueDictionary values,
        RouteDirection routeDirection)
    {
        if (routeKey == null)
        {
            throw new ArgumentNullException(nameof(routeKey));
        }

        if (values == null)
        {
            throw new ArgumentNullException(nameof(values));
        }

        if (!values.TryGetValue(routeKey, out var routeValue) || routeValue == null)
        {
            return false;
        }
        if (routeValue is ulong)
        {
            return true;
        }

        var valueString = Convert.ToString(routeValue, CultureInfo.InvariantCulture);
        var parsed = ulong.TryParse(valueString, out var value);
        if (parsed)
        {
            values[routeKey] = value;
        }

        return parsed;
    }
}
```

Register:
```
builder.Services.Configure<RouteOptions>(options =>
{
    options.ConstraintMap.Add(UlongRouteConstraint.UlongRouteConstraintName, typeof(UlongRouteConstraint));
});
```

Use:
```
@page "/user/servers/{ServerId:ulong}/{ProductId:ulong}/details"

...


[Parameter, EditorRequired]
public ulong ServerId { get; set; }

[Parameter, EditorRequired]
public ulong ProductId { get; set; }
```

**Comment by AndreyZ1** [comment]
> It looked promising, and I am trying this approach (works in server-side WebApi), but in my Blazor WASM I am getting an exception:
> 
> > Unhandled exception rendering component: The constraint entry 'parameter1' - 'ulid' on the route '/path/{parameter1:ulid}/{parameter2:ulid}' could not be resolved by the constraint resolver of type 'DefaultInlineConstraintResolver'.
> > System.InvalidOperationException: The constraint entry 'parameter1' - 'ulid' on the route '/path/{parameter1:ulid}/{parameter2:ulid}' could not be resolved by the constraint resolver of type 'DefaultInlineConstraintResolver'.
> 
> `@page "/path/{parameter1:ulid}/{parameter2:ulid}"`
> 
> Even though I have:
> 
> ```
> public class UlidRouteConstraint : IRouteConstraint
> {
>     public const string ConstraintName = "ulid";
> 
>     public bool Match(HttpContext? httpContext, IRouter? route, string routeKey, RouteValueDictionary values, RouteDirection routeDirection)
>     {
>         if (values.TryGetValue(routeKey, out var value) && value is string stringValue)
>         {
>             return Ulid.TryParse(stringValue, out _);
>         }
>         return false;
>     }
> }
> ```
> 
> and
> 
> ```
> builder.Services.Configure<RouteOptions>(options =>
> {
>     options.ConstraintMap.Add(UlidRouteConstraint.ConstraintName, typeof(UlidRouteConstraint));
> });
> ```

are you sure you registered RouteOptions for WASM project as well?

**Comment by zorgoz** [comment]
Absolutely, That's from my actual code. I have put a breakpoint at the `options.ConstraintMap.Add` and it is never hit.

**Comment by AndreyZ1** [comment]
> Absolutely, That's from my actual code. I have put a breakpoint at the `options.ConstraintMap.Add` and it is never hit.

Can not help you here.
As long as server and client knows how to handle custom type - it should work

**Comment by zorgoz** [comment]
> > Absolutely, That's from my actual code. I have put a breakpoint at the `options.ConstraintMap.Add` and it is never hit.
> 
> Can not help you here. As long as server and client knows how to handle custom type - it should work

You can confirm that you are using in a WebAssembly standalone (not interactive WebAssembly ) and it is working for you?

**Comment by AndreyZ1** [comment]
> > > Absolutely, That's from my actual code. I have put a breakpoint at the `options.ConstraintMap.Add` and it is never hit.
> > 
> > 
> > Can not help you here. As long as server and client knows how to handle custom type - it should work
> 
> You can confirm that you are using in a Blazor WebAssembly standalone (not interactive WebAssembly ) and it is working for you?

Currently I don't have solution running anymore, but I used it in WASM client with .net Core app as a server.

**Comment by scss-sconnors** [comment]
Above solution doesnt seem to work if setting the rendermode on `Routes.razor` to `InteractiveServer` and adding additional assemblies via the router component

e.g.

`<head>
    @* ... *@
    <HeadOutlet @rendermode="InteractiveServer" />
</head>
<body>
    <Routes @rendermode="InteractiveServer" />
    @* ... *@
</body>`


`<Router
    AppAssembly="..."
    AdditionalAssemblies="[ typeof(BlazorSample.Client._Imports).Assembly ]">
    ...
</Router>`

It seems in interactive mode, even with constraints defined in program.cs, the router component seems to set up a brand new DefaultInlineConstraintResolver, discarding our custom-mapped constaints.

Is there anyway around this huge limitation? I just want the app to display a loading spinner instead of pre-rendering the entire app just for it to be thrown away seconds later.

**Comment by amdavie** [comment]
I am also really needing the ability to define custom route constraints for Blazor Server apps.

The approach mentioned in this thread of creating a custom `IRouteConstraint` implementation and configuring it via `RouteOptions.ConstraintMap` is not compatible with the `Router` component due to `RouteTableFactory` not using the `RouteOptions` configured with the DI container but instead creating a new instance. Therefore, the `DefaultInlineConstraintResolver` will not recognize any custom constraint that is mapped.

https://github.com/dotnet/aspnetcore/blob/5f1f2ce6ece7ec84f68c848dc1a6f56209c1c0da/src/Components/Components/src/Routing/RouteTableFactory.cs#L116-L125

The `IServiceProvider` is already passed into the static `RouteTableFactory.Create` method so it seems like it should be trivial to resolve the `RouteOptions` instead. Is there something about the `Router` that would make it incompatible with the `RouteOptions` registered with DI?
